hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HRegionServer.java

```diff
@@ -1659,6 +1662,43 @@ public RegionLoad createRegionLoad(final String encodedRegionName) throws IOExce
     return r != null ? createRegionLoad(r, null, null) : null;
   }
 
+  public Map<String, Integer> uncacheStaleBlocks() {
+    Map<String, Pair<String, Long>> fullyCachedFiles =
+      this.getBlockCache().flatMap(BlockCache::getFullyCachedFiles).orElse(Collections.emptyMap());
+    Map<String, Integer> evictedFilesWithStaleBlocks = new ConcurrentHashMap<>();
+
+    ExecutorService executor = Executors.newFixedThreadPool(6);
+
+    List<Callable<Void>> tasks = new ArrayList<>();
+
+    fullyCachedFiles.forEach((fileName, value) -> {
+      Callable<Void> task = () -> {
+        HRegion regionOnServer = getRegion(value.getFirst());
+        int blocksEvicted = (regionOnServer == null || !regionOnServer.isAvailable())
+          ? this.getBlockCache().get().evictBlocksByHfileName(fileName)
+          : 0;
+        evictedFilesWithStaleBlocks.put(fileName, blocksEvicted);
+        LOG.info(
+          "Uncached {} blocks belonging to the file {} as the region {} "
+            + "is not served by the region server {} anymore.",
+          blocksEvicted, fileName, value.getFirst(), this.getServerName());
+        return null;
+      };
+      tasks.add(task);
+    });
+
+    try {
+      executor.invokeAll(tasks);
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      LOG.error("Thread interrupted while processing tasks for uncaching stale blocks: {}",
+        e.getMessage());
+    } finally {
+      executor.shutdown();
+    }
+    return evictedFilesWithStaleBlocks;
```

Comment:
We should think about an alternative result here, as this map is effectively being updated by the background task, which can be confusing for the caller. I would rather make this as void, or a boolean, returning true indicating that the task is submitted and is running in the background.